specs/packagemanager/PackageManagement.md

```diff
@@ -635,6 +644,66 @@ void CheckAndFixPackageVolume(string packageStorePath)
 }
 ```
 
+## 4.9. IsPackageProvisioned()
+
+Fabrikam app installing Contoso's Muffin and Waffle packages for all users if necessary, and with explicit user confirmation before the installation.
+
+```c#
+void Install()
+{
+    var packageSet = new PackageSet() {
+        Items = { new PackageSetItem() { PackageFamilyName = "contoso.muffin_1234567890abc",
+                                         MinVersion = ToVersion(1, 2, 3, 4),
+                                         PackageUri = new Uri("c:\\contoso\\muffin-1.2.3.4.msix") },
+                { new PackageSetItem() { PackageFamilyName = "contoso.waffle_1234567890abc",
+                                         MinVersion = ToVersion(2, 4, 6, 8),
+                                         PackageUri = new Uri("https://contoso.com/waffle-2.4.6.8.msix") } };
+
+    var packageDeploymentManager = PackageDeploymentManager.GetDefault();
+    if (!packageDeploymentManager.IsPackageSetProvisioned(packageSet))
```

Comment:
Does this check only that the specified package families have been provisioned (at any version), or does it also check the version numbers of each package in the set?

If the currently provisioned version is *newer* than the one requested, does Is\*Provisioned return "true" (it's even better than you dreamed!) or "false" (the exact version you are asking about is not provisioned)?

Reply:
Related is there an exact version check available sometimes packages break compatibility between versions

Reply:
>chrilaMSFT is there an exact version check available sometimes packages break compatibility between versions

Provisioning is based on Package Family so 'exact version' is meaningless.

If you're interested in what's staged on the machine you can do an exact match via `pkgmgr.FindPackage(pkgfullname)` and if you get Package object it's found. I think you get a specific error condition if not found (`HRESULT_FROM_WIN32(ERROR_NOT_FOUND)`?) rather than the usual `null` as PackageManager is one of the oldest WinRT APIs (early Win8 days) and doesn't follow a lot of common practices by today's standards. That's one of the reasons for `PackageDeploymentManager`, as a "If you created PackageManager today...".

Similarly you can explicitly check if a package is registered for a user via `pkgmgr.FindPackageForUser("", pkgfullname)`, with similar caveats about detecting not-found.

You can also do these via non-WinRT APIs e.g. [GetStagedPackageOrigin()](https://learn.microsoft.com/windows/win32/api/appmodel/nf-appmodel-getstagedpackageorigin), [FindPackagesByPackageFamily()](https://learn.microsoft.com/windows/win32/api/appmodel/nf-appmodel-findpackagesbypackagefamily) and other APIs (usually in appmodel.h).

`PackageDeploymentManager` provides a superset of most things in `PackageManager` but notably no `FindPackage*()` methods. Not yet. Doing that right is more involved than it sounds and improved 'query' APIs weren't as urgent a need as 'do' methods e.g. `EnsurePackageSetIsReady()`. I expect we'll get there one day, but more urgent matters first.

Reply:
> Does this check only that the specified package families have been provisioned (at any version), or does it also check the version numbers of each package in the set?

The former. Provisioning is based on Package Family. You don't provision Foo-v1, you provision Foo. If you provision Foo-v1 via `DISM` or the powershell `Add-AppxProvisionedPackage` then you're really staging Foo-v1 and provisioning Foo.

The sample implies it but shouldn't.

The table in 3.1. API Structure erroneously states IsProvisioned supports PackageFullName. 

RECOMMEND: Improve the sample. Also fix the table (as noted above).

Reply:
Particularly misleading since there is a MinVersion that is apparently ignored? Or is that what you are referring to by "the sample implies it but shouldn't"?

But even so, the fact that there's a PackageSet variant, and PackageSetItems have a MinVersion, the existence of a MinVersion is unavoidable.

Reply:
I am confused there is a package version and url for where to install the package but we don't check the version so url is not used if any package in the family is installed.  We have apis that can detect the version individually and can do the same logic.  Why do we need the new API.  Why not just write a helper function that is open source.